fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,380 @@
+	    refreshByCacheInit("TenantEntryCacheRefreshInit", metrics),
+	    refreshByCacheMiss("TenantEntryCacheRefreshMiss", metrics),
+	    numRefreshes("TenantEntryCacheNumRefreshes", metrics) {
+		TraceEvent("TenantEntryCacheCreated", uid);
+	}
+
+	TenantEntryCache(Database db, UID id, TenantEntryCachePayloadFunc<T> fn)
+	  : uid(id), db(db), createPayloadFunc(fn), refreshMode(TenantEntryCacheRefreshMode::PERIODIC_TASK),
+	    metrics("TenantEntryCacheMetrics", uid.toString()), hits("TenantEntryCacheHits", metrics),
+	    misses("TenantEntryCacheMisses", metrics), refreshByCacheInit("TenantEntryCacheRefreshInit", metrics),
+	    refreshByCacheMiss("TenantEntryCacheRefreshMiss", metrics),
+	    numRefreshes("TenantEntryCacheNumRefreshes", metrics) {
+		TraceEvent("TenantEntryCacheCreated", uid);
+	}
+
+	TenantEntryCache(Database db, UID id, TenantEntryCachePayloadFunc<T> fn, TenantEntryCacheRefreshMode mode)
+	  : uid(id), db(db), createPayloadFunc(fn), refreshMode(mode), metrics("TenantEntryCacheMetrics", uid.toString()),
+	    hits("TenantEntryCacheHits", metrics), misses("TenantEntryCacheMisses", metrics),
+	    refreshByCacheInit("TenantEntryCacheRefreshInit", metrics),
+	    refreshByCacheMiss("TenantEntryCacheRefreshMiss", metrics),
+	    numRefreshes("TenantEntryCacheNumRefreshes", metrics) {
+		TraceEvent("TenantEntryCacheCreated", uid);
+	}
+
+	Future<Void> init() {
+		TraceEvent(SevInfo, "TenantEntryCacheInit", uid).log();
+
+		Future<Void> f = refreshImpl(this, TenantEntryCacheRefreshReason::INIT);
+
+		// Launch reaper task to periodically refresh cache by scanning database KeyRange
+		TenantEntryCacheRefreshReason reason = TenantEntryCacheRefreshReason::PERIODIC_TASK;
+		if (refreshMode == TenantEntryCacheRefreshMode::PERIODIC_TASK) {
+			refresher = recurringAsync([&, reason]() { return refresh(reason); },
+			                           SERVER_KNOBS->TENANT_CACHE_LIST_REFRESH_INTERVAL, /* interval */
+			                           true, /* absoluteIntervalDelay */
+			                           SERVER_KNOBS->TENANT_CACHE_LIST_REFRESH_INTERVAL, /* intialDelay */
+			                           TaskPriority::Worker);
+		}
+
+		return f;
+	}
+
+	Database getDatabase() const { return db; }
+	UID id() const { return uid; }
+
+	void clear() {
+		mapByTenantId.clear();
+		mapByTenantName.clear();
+	}
+
+	Future<Void> removeEntryById(int64_t tenantId, bool refreshCache = false) {
+		return removeEntryInt(tenantId, Optional<KeyRef>(), Optional<TenantName>(), refreshCache);
+	}
+	Future<Void> removeEntryByPrefix(KeyRef tenantPrefix, bool refreshCache = false) {
+		return removeEntryInt(Optional<int64_t>(), tenantPrefix, Optional<TenantName>(), refreshCache);
+	}
+	Future<Void> removeEntryByName(TenantName tenantName, bool refreshCache = false) {
+		return removeEntryInt(Optional<int64_t>(), Optional<KeyRef>(), tenantName, refreshCache);
+	}
+
+	void put(const TenantNameEntryPair& pair) {
```

Comment:
Should this be marked private? The contract assumed here matches what I would expect for our internal use of the function, but an external use of the function should probably not be assumed to have those properties. In particular, a user could attempt to put an updated entry where either the name or ID has changed (due to a rename or a delete/create cycle) that could violate the consistency of the two maps.

Reply:
> that could violate the consistency of the two maps.

I think allowing client to update entry seems like a useful functionality of a cache; `rename` or create/delete` cycle are valid points, one way to ensure map consistency is first remove entries from both maps by looking up using `TenantId` and insert the new entry later ; assumption here `TenantId` is not-reused and unique.

Does this sounds reasonable?

Reply:
That seems fine too, I wasn't sure if you intended for this to be something that could be updated in that way. Removing and inserting the entries is fine, are you planning that `put` would do both of those steps?

There are two different cases that immediately come to mind:

1. The tenant already exists with the same ID but has a different name (i.e. it was renamed). Then you could search for a tenant with a matching ID and use the returned name to delete it from the name map.
2. The tenant already exists with the same name but has a different ID (i.e. you deleted and recreated the tenant). In this case, I think you need to search the name map to see if it exists and then remove the entry with the returned ID.